lib/BloqadeExpr/src/types.jl

```diff
@@ -281,6 +281,66 @@ end
 
 SumOfZ(n::Int) = SumOfZ(n, 1)
 
+
+RabiTypes = Union{Nothing,SumOfX,SumOfXPhase}
+DetuningTypes = Union{Nothing,SumOfN}
+
+struct RydbergHamiltonian{RabiType <: RabiTypes, DetuningType <: DetuningTypes} <: AbstractTerm
+    rydberg_term::RydInteract
+    rabi_term::RabiType
+    detuning_term::DetuningType
+end
+
+@inline add_terms(h::RydbergHamiltonian{Nothing,Nothing}) = YaoBlocks.Optimise.simplify(h.rydberg_term)
```

Comment:
just overload `to_basicblocks` for this as an expression transform, and use `simplify` to do the simplification. I think `if else` style is preferred here so we can get rid of the dynamic dispatch.

Reply:
also note this won't work if `RydbergHamiltonian` is inside other expression, but

https://github.com/QuEraComputing/Bloqade.jl/blob/master/lib/BloqadeExpr/src/mat.jl#L25

will work